src/history/hs.c

```diff
@@ -1236,121 +1234,185 @@ __hs_delete_key_from_pos(
-            F_SET(data_cursor, WT_CURSOR_RAW_OK);
-        } else {
-            WT_ERR(__wt_compare(session, NULL, hs_key, prev_hs_key, &cmp));
-            if (cmp == 0)
-                continue;
-        }
-        WT_ERR(__wt_buf_set(session, prev_hs_key, hs_key->data, hs_key->size));
-        prev_btree_id = btree_id;
-
-        /* Re-purpose the previous-key buffer on error, safe because we're about to error out. */
-        data_cursor->set_key(data_cursor, hs_key);
-        if ((ret = data_cursor->search(data_cursor)) == WT_NOTFOUND)
-            WT_ERR_MSG(session, ret,
-              "In %s, the associated history store key %s was not found in the data store", uri_itr,
-              __wt_buf_set_printable(session, hs_key->data, hs_key->size, prev_hs_key));
+        WT_ERR(__wt_compare(session, NULL, hs_key, prev_hs_key, &cmp));
+        if (cmp == 0)
+            continue;
+        WT_WITH_PAGE_INDEX(session, ret = __wt_row_search(cbt, hs_key, false, NULL, false, NULL));
         WT_ERR(ret);
+
+        found = cbt->compare == 0;
+        WT_ERR(__cursor_reset(cbt));
+
+        if (!found)
+            WT_ERR_MSG(session, WT_PANIC,
+              "the associated history store key %s was not found in the data store %s",
+              __wt_buf_set_printable(session, hs_key->data, hs_key->size, prev_hs_key),
+              session->dhandle->name);
+
+        /* Swap current/previous buffers. */
+        tmp = hs_key;
+        hs_key = prev_hs_key;
+        prev_hs_key = tmp;
     }
-    WT_ERR_NOTFOUND_OK(ret, false);
+    WT_ERR_NOTFOUND_OK(ret, true);
 err:
-    if (data_cursor != NULL)
+    __wt_scr_free(session, &hs_key);
+    __wt_scr_free(session, &prev_hs_key);
+    return (ret);
+}
+
+/*
+ * __wt_history_store_verify_one --
+ *     Verify the history store for the btree that is set up in this session. This must be called
+ *     when we are known to have exclusive access to the btree.
+ */
+int
+__wt_history_store_verify_one(WT_SESSION_IMPL *session)
+{
+    WT_CURSOR *cursor;
+    WT_CURSOR_BTREE cbt;
+    WT_DECL_RET;
+    WT_ITEM hs_key;
+    uint32_t btree_id;
+    int exact;
+
+    cursor = session->hs_cursor;
+    btree_id = S2BT(session)->id;
+
+    /*
+     * We are required to position the history store cursor. Set it to the first record of our btree
+     * in the history store.
+     */
+    memset(&hs_key, 0, sizeof(hs_key));
+    cursor->set_key(cursor, btree_id, &hs_key, 0, 0, 0, 0);
+    ret = cursor->search_near(cursor, &exact);
+    if (ret == 0 && exact < 0)
+        ret = cursor->next(cursor);
+    WT_ERR_NOTFOUND_OK(ret, true);
+
+    /* If we positioned the cursor there is something to verify. */
+    if (ret != WT_NOTFOUND) {
+        __wt_btcur_init(session, &cbt);
+        __wt_btcur_open(&cbt);
+        ret = __verify_history_store_id(session, &cbt, btree_id);
+        WT_TRET(__wt_btcur_close(&cbt, false));
+        WT_ERR_NOTFOUND_OK(ret, false);
+    }
+    ret = 0;
+err:
```

Comment:
This might be a little simpler as:
```
    ret = cursor->search_near(cursor, &exact);
    if (ret == 0 && exact < 0)  
        ret = cursor->next(cursor);
        
    /* If we positioned the cursor there is something to verify. */
    if (ret == 0) {
        __wt_btcur_init(session, &cbt);
        __wt_btcur_open(&cbt);
        ret = __verify_history_store_id(session, &cbt, btree_id);
        WT_TRET(__wt_btcur_close(&cbt, false));
    }
    return (ret == WT_NOTFOUND ? 0 : ret);
```
But I wouldn't argue about it.

Reply:
Fixed.